### ISSUE TITLE
cumsum axis bug

### ISSUE THREAD

**Issue by SteveDiamond** [cvxpy/cvxpy#1678]
**Describe the bug**
``cumsum(x, axis=1)`` breaks when ``x`` has shape ``(n, 1)``. Similarly, ``cumsum(x, axis=0)`` breaks when ``x`` has shape ``(1, n)``. This bug should be easy to fix by adding special cases to https://github.com/cvxpy/cvxpy/blob/master/cvxpy/reductions/eliminate_pwl/atom_canonicalizers/cumsum_canon.py

**To Reproduce**
```
import cvxpy as cp

n = 5
axis = 1

if axis == 0:
  x = cp.Variable((1, n))
  expr = cp.cumsum(x, axis=0)

  prob = cp.Problem(cp.Minimize(0),
                   [expr == 1])
  prob.solve()
  print(x.value)
elif axis == 1:
  x = cp.Variable((n, 1))
  expr = cp.cumsum(x, axis=1)

  prob = cp.Problem(cp.Minimize(0),
                   [expr == 1])
  prob.solve()
  print(x.value)
```

**Expected behavior**
The problem should solve without crashing :P

**Output**
```
Traceback (most recent call last):
  File "/Users/sdiamond/Projects/tests/cumsum_test.py", line 20, in <module>
    prob.solve()
  File "/Users/sdiamond/opt/anaconda3/envs/cvxpy/lib/python3.9/site-packages/cvxpy/problems/problem.py", line 473, in solve
    return solve_func(self, *args, **kwargs)
  File "/Users/sdiamond/opt/anaconda3/envs/cvxpy/lib/python3.9/site-packages/cvxpy/problems/problem.py", line 966, in _solve
    data, solving_chain, inverse_data = self.get_problem_data(
  File "/Users/sdiamond/opt/anaconda3/envs/cvxpy/lib/python3.9/site-packages/cvxpy/problems/problem.py", line 627, in get_problem_data
    data, inverse_data = solving_chain.apply(self, verbose)
  File "/Users/sdiamond/opt/anaconda3/envs/cvxpy/lib/python3.9/site-packages/cvxpy/reductions/chain.py", line 76, in apply
    problem, inv = r.apply(problem)
  File "/Users/sdiamond/opt/anaconda3/envs/cvxpy/lib/python3.9/site-packages/cvxpy/reductions/dcp2cone/dcp2cone.py", line 43, in apply
    return super(Dcp2Cone, self).apply(problem)
  File "/Users/sdiamond/opt/anaconda3/envs/cvxpy/lib/python3.9/site-packages/cvxpy/reductions/canonicalization.py", line 67, in apply
    canon_constr, aux_constr = self.canonicalize_tree(
  File "/Users/sdiamond/opt/anaconda3/envs/cvxpy/lib/python3.9/site-packages/cvxpy/reductions/canonicalization.py", line 99, in canonicalize_tree
    canon_arg, c = self.canonicalize_tree(arg)
  File "/Users/sdiamond/opt/anaconda3/envs/cvxpy/lib/python3.9/site-packages/cvxpy/reductions/canonicalization.py", line 102, in canonicalize_tree
    canon_expr, c = self.canonicalize_expr(expr, canon_args)
  File "/Users/sdiamond/opt/anaconda3/envs/cvxpy/lib/python3.9/site-packages/cvxpy/reductions/canonicalization.py", line 113, in canonicalize_expr
    return self.canon_methods[type(expr)](expr, args)
  File "/Users/sdiamond/opt/anaconda3/envs/cvxpy/lib/python3.9/site-packages/cvxpy/reductions/eliminate_pwl/atom_canonicalizers/cumsum_canon.py", line 31, in cumsum_canon
    constr = [X[:, 1:] == Y[:, 1:] - Y[:, :-1], Y[:, 0] == X[:, 0]]
  File "/Users/sdiamond/opt/anaconda3/envs/cvxpy/lib/python3.9/site-packages/cvxpy/expressions/expression.py", line 49, in cast_op
    return binary_op(self, other)
  File "/Users/sdiamond/opt/anaconda3/envs/cvxpy/lib/python3.9/site-packages/cvxpy/expressions/expression.py", line 658, in __eq__
    return Equality(self, other)
  File "/Users/sdiamond/opt/anaconda3/envs/cvxpy/lib/python3.9/site-packages/cvxpy/constraints/zero.py", line 104, in __init__
    super(Equality, self).__init__([lhs, rhs], constr_id)
  File "/Users/sdiamond/opt/anaconda3/envs/cvxpy/lib/python3.9/site-packages/cvxpy/constraints/constraint.py", line 51, in __init__
    self._construct_dual_variables(args)
  File "/Users/sdiamond/opt/anaconda3/envs/cvxpy/lib/python3.9/site-packages/cvxpy/constraints/zero.py", line 118, in _construct_dual_variables
    super(Equality, self)._construct_dual_variables([self._expr])
  File "/Users/sdiamond/opt/anaconda3/envs/cvxpy/lib/python3.9/site-packages/cvxpy/constraints/constraint.py", line 66, in _construct_dual_variables
    self.dual_variables = [cvxtypes.variable()(arg.shape) for arg in args]
  File "/Users/sdiamond/opt/anaconda3/envs/cvxpy/lib/python3.9/site-packages/cvxpy/constraints/constraint.py", line 66, in <listcomp>
    self.dual_variables = [cvxtypes.variable()(arg.shape) for arg in args]
  File "/Users/sdiamond/opt/anaconda3/envs/cvxpy/lib/python3.9/site-packages/cvxpy/expressions/variable.py", line 86, in __init__
    super(Variable, self).__init__(shape, **kwargs)
  File "/Users/sdiamond/opt/anaconda3/envs/cvxpy/lib/python3.9/site-packages/cvxpy/expressions/leaf.py", line 108, in __init__
    raise ValueError("Invalid dimensions %s." % (shape,))
ValueError: Invalid dimensions (5, 0).
```

**Version**
 - OS: macOS 12
 - CVXPY Version: 1.1.18

**Additional context**
N/A

**Comment by parthb83** [comment]
Hey @SteveDiamond! So I added the following case and it seems to run the sample code given by you without showing any error and giving the correct output. Do tell if it is correct or not and mention any change that it may require.
The link to PR: #1680.

```
if axis == 0:
        if (expr.shape[0] == 1):
            constr = [Y[0] == X[0]]
        else:
            constr = [X[1:] == Y[1:] - Y[:-1], Y[0] == X[0]]
else:
        if (expr.shape[1] == 1):
            constr = [Y[:, 0] == X[:, 0]]
        else:
            constr = [X[:, 1:] == Y[:, 1:] - Y[:, :-1], Y[:, 0] == X[:, 0]]```

**Comment by rileyjmurray** [comment]
@SteveDiamond @parthb83 this wasn't fully resolved. An analogous bug was in ``cummax``. I've implemented the same fix on the PR for CVXPY 1.1.19.  @parthb83 please make another PR which incorporates those fixes (and updates tests for cumsum).

**Comment by parthb83** [comment]
Hey @rileyjmurray! I've made the changes suggested by you and opened the PR #1683 for the same. Kindly review the PR.

**Comment by rileyjmurray** [comment]
Resolved by #1683 :)
